src/locutus/model/terminology.py

```diff
@@ -649,6 +652,131 @@ def replace_preferred_terminology(self, editor, preferred_terminology):
+
+                    # Handle different input types
+                    if type == "mapping_conversations":
+                        schema = UserInput.MappingConversations._Schema()
+                        input_data = existing_data.get("mapping_conversations", [])
+                    elif type == "mapping_votes":
+                        schema = UserInput.MappingVotes._Schema()
+                        input_data = existing_data.get("mapping_votes", {})
+                    else:
+                        return {"message": "Invalid input type specified."}, 400
+
+                    # Serialize the data using the appropriate schema
+                    serialized_data = schema.dump({type: input_data})
+
+                    return {"Terminology": id,
+                            "code": code,
+                            type: serialized_data[type]}
+
+                else:
+                    # Return an empty response if no input exists
+                    return {"Terminology": id, "code": code, "message": "No user input for this mapping."}
+            
+            except Exception as e:
+                return (f"An error occurred while retrieving user input for {id} {resource_type} - {code}: {e}"), 500
+
+
+    
+    def create_or_replace_user_input(self, resource_type, collection_type, id, code, type, user_input):
+        """
+        Creates or replaces a document in the 'user_input' sub-collection data
+        for a user.
+        """
+        try:
+            doc_ref = persistence().collection(resource_type).document(id) \
+                .collection(collection_type).document(code)
+
+            # Fetch existing data for the document if it exists
+            doc_snapshot = doc_ref.get()
+            existing_data = doc_snapshot.to_dict() if doc_snapshot.exists else {}
+
+            # Get user_id to identify existing data for the user.
+            user_id = SessionManager.create_session_user_id()
+            if 'error' in user_id:
+                raise ValueError("User not logged in")
+        
+            # Initialize type structure if it doesn't exist. Debug: If put 
+            # request does not add data the result will be an empty list or dict.
+            if type not in existing_data:
+                if type == "mapping_votes":
+                    existing_data[type] = {}
+                elif type == "mapping_conversations":
+                    existing_data[type] = []
+
+            # Update or append the user input based on type. Updates existing_data.
+            self.update_or_append_input(existing_data[type], user_id, user_input)
+            
+            # Update the Firestore document with the new data.
+            doc_ref.set(existing_data)
+
+        except Exception as e:
+            return (f"An error occurred while retrieving user input for {id} \
+                    {resource_type} - {code}: {e}"), 500
+
+
+    def update_or_append_input(self, input_data, user_id, user_input):

```

Comment:
For this one, since each of those two user input classes behaves differently, you probably want to implement the function in each one.